src/tools/unstable-book-gen/src/main.rs

```diff
@@ -11,13 +13,121 @@ use tidy::unstable_book::{
+use rustdoc_json_types::{Crate, Id};
+use syn::{parse::Parser, Ident, Lit, Meta, NestedMeta};
+
+// (item path, doc url)
+fn full_path(krate: &Crate, item: &Id) -> Option<(String, String)> {
+    let item_summary = krate.paths.get(item)?;
+    let kind = &item_summary.kind;
+    let kind_str = serde_json::to_string(kind).ok()?;
+    let mut url = String::from("https://doc.rust-lang.org/nightly/");
+    let mut iter = item_summary.path.iter();
+    iter.next_back();
+    url.push_str(&iter.cloned().collect::<Vec<_>>().join("/"));
+    url.push('/');
+    url.push_str(kind_str.trim_matches('"'));
+    url.push('.');
+    url.push_str(item_summary.path.last().unwrap());
+    url.push_str(".html");
+    Some((item_summary.path.join("::"), url))
+}
+
+fn is_ident(ident: &Ident, name: &str) -> bool {
+    *ident == Ident::new(name, ident.span())
+}
+
+/// Returns an unstable feature -> (item path, doc url) mapping.
+pub fn load_rustdoc_json_metadata(doc_dir: &Path) -> HashMap<String, Vec<(String, String)>> {
+    let mut all_items = HashMap::new();
+
+    for file in fs::read_dir(doc_dir).expect("failed to list files in directory") {
+        let entry = file.expect("failed to list file in directory");
+        let file = fs::File::open(entry.path()).expect("failed to open file");
+        let krate: Crate =
+            serde_json::from_reader(BufReader::new(file)).expect("failed to parse JSON docs");
+
+        let mut crate_items = HashMap::new();
+        for (id, item) in &krate.index {
+            if item.name.is_none() {
+                continue;
+            }
+            let unstable_feature = item.attrs.iter().find_map(|attr: &String| {
+                let Ok(parsed) = syn::Attribute::parse_outer.parse_str(attr).map(|mut v| v.swap_remove(0)) else {return None};
+
+                // Make sure this is an `unstable` attribute.
+                if !is_ident(parsed.path.get_ident()?, "unstable") {
+                    return None;
+                }
+
+                // Given `#[unstable(feature = "xyz")]`, return `(feature = "xyz")`.
+                let list = match parsed.parse_meta() {
+                    Ok(Meta::List(list)) => list,
+                    _ => return None,
+                };
+
+                // Given a `NestedMeta` like `feature = "xyz"`, returns `xyz`.
+                let get_feature_name = |nested: &_| {
+                    match nested {
+                        NestedMeta::Meta(Meta::NameValue(name_value)) => {
+                            if !is_ident(name_value.path.get_ident()?, "feature") {
+                                return None;
+                            }
+                            match &name_value.lit {
+                                Lit::Str(s) => Some(s.value()),
+                                _ => None,
+                            }
+                        }
+                        _ => None,
+                    }
+                };
```

Comment:
This can be defined outside of the loop.

Also, we can reduce the scopes with the following way
```rs
// Given a `NestedMeta` like `feature = "xyz"`, returns `xyz`.
let get_feature_name = |nested: &_| match nested {
    NestedMeta::Meta(Meta::NameValue(name_value)) => {
        if !is_ident(name_value.path.get_ident()?, "feature") {
            return None;
        }
        match &name_value.lit {
            Lit::Str(s) => Some(s.value()),
            _ => None,
        }
    }
    _ => None,
};
```